src/liballoc/collections/binary_heap.rs

```diff
@@ -1115,6 +1178,34 @@ impl<T> ExactSizeIterator for IntoIter<T> {
 #[stable(feature = "fused", since = "1.26.0")]
 impl<T> FusedIterator for IntoIter<T> {}
 
+#[unstable(feature = "binary_heap_into_iter_ordered", issue = "59278")]
+#[derive(Clone, Debug)]
+pub struct IntoIterOrdered<T> {
+    inner: BinaryHeap<T>,
+}
+
+#[unstable(feature = "binary_heap_into_iter_ordered", issue = "59278")]
+impl<T: Ord> Iterator for IntoIterOrdered<T> {
+    type Item = T;
+
+    #[inline]
+    fn next(&mut self) -> Option<T> {
+        self.inner.pop()
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let exact = self.inner.len();
+        (exact, Some(exact))
+    }
+}
+
+#[unstable(feature = "binary_heap_into_iter_ordered", issue = "59278")]
+impl<T: Ord> ExactSizeIterator for IntoIterOrdered<T> {}
+
+#[unstable(feature = "binary_heap_into_iter_ordered", issue = "59278")]
+impl<T: Ord> FusedIterator for IntoIterOrdered<T> {}
```

Comment:
I think you can add `TrustedLen` for both.

Reply:
I'll add `TrustedLen` for the two new iterators.

I'm wondering if other existing iterators should implement `TrustedLen` for consistency. 
* `binary_heap::IntoIter`
  * Underlying `vec::IntoIter` implements `TrustedLen`
  * Clearly `binary_heap::IntoIter` should implement `TrustedLen`, too.
* `binary_heap::Drain`
  * Underlying `vec::Drain` does not implement `TrustedLen`.
  * It is not clear whether it should implement `TrustedLen`.
* `vec::Drain`
  * It is not clear whether it should implement `TrustedLen`.

Please advise.

Reply:
They should implement `TrustedLen` if and only if it would be sound to do so. However, I think it's best to make a follow-up issue and treat with them there.

Reply:
I see, I go with conservative way.

Reply:
`TrustedLen` was added.